design/design-doc-e2e-testing.md

```diff
@@ -0,0 +1,809 @@
+```go
+func InitializeScenario(ctx *godog.ScenarioContext) {
+	ctx.Before(func(ctx context.Context, scenario *godog.Scenario) (context.Context, error) {
+		sc := &scenarioContext{
+			Cluster: &cluster{
+				cli: "kubectl",
+			},
+			Namespace: fmt.Sprintf("test-%s", scenario.Id),
+		}
+		if err := sc.Cluster.createNamespace(sc.Namespace); err != nil {
+			return ctx, err
+		}
+		ctx = context.WithValue(ctx, scenarioContextKey, sc)
+		return ctx, nil
+	})
+	ctx.Step(`^claim becomes synchronized and ready$`, claimBecomesSynchronizedAndReady)
+	ctx.Step(`^claim composite resource becomes synchronized and ready$`, claimCompositeResourceBecomesSynchronizedAndReady)
+	ctx.Step(`^claim gets deployed$`, claimGetsDeployed)
+	ctx.Step(`^composed managed resources become ready and synchronized$`, composedManagedResourcesBecomeReadyAndSynchronized)
+	ctx.Step(`^CompositeResourceDefinition is present$`, clusterScopedResourceIsPresent)
+	ctx.Step(`^Composition is present$`, clusterScopedResourceIsPresent)
+	ctx.Step(`^Configuration is applied$`, configurationGetsDeployed)
+	ctx.Step(`^configuration is marked as installed and healthy$`, configurationMarkedAsInstalledAndHealthy)
+	ctx.Step(`^Crossplane is running in cluster$`, crossplaneIsRunningInCluster)
+	ctx.Step(`^provider (\S+) does not get installed$`, providerNotInstalled)
+	ctx.Step(`^provider (\S+) is marked as installed and healthy$`, providerMarkedAsInstalledAndHealthy)
+	ctx.Step(`^provider (\S+) is running in cluster$`, providerGetsInstalled)
+}
+```
+Additional hooks and context are registered as well.
+
+Finally, the tests are triggered as regular go tests:
+
+```go
+var opts = godog.Options{
+	Output: colors.Colored(os.Stdout),
+	Format: "pretty",
+}
+
+func init() {
+	godog.BindFlags("godog.", flag.CommandLine, &opts)
+}
+
+func TestE2E(t *testing.T) {
+	o := opts
+	o.TestingT = t
+
+	status := godog.TestSuite{
+		ScenarioInitializer: InitializeScenario,
+		Options:             &o,
+	}.Run()
+
+	if status == 2 {
+		t.SkipNow()
+	}
+
+	if status != 0 {
+		t.Fatal("non-zero status returned, failed to run feature tests")
+	}
+}
+```
+The existing e2e tests have been converted to Cucumber scenarios, and they could be executed with:
+
+```shell
+make build e2e
+```
+
+When it comes to choosing a provider to run tests against, we currently have the
+[NOP provider] used by the current e2e tests. An alternative provider could be
+[provider-dummy]. We believe that most of the tests could be even written so
+that they do not require any provider to be installed, i.e. it would be enough
+to install some carefully crafted CRDs, and create XRDs and claims against them.
+Such a strategy would make required setups even lighter and tests would
+potentially run faster.
+
+### Blackbox Testing
+
+Steps implementation communicates with the cluster only by using [kubectl CLI].
+In this way we can be sure that users are able to reproduce tests either
+manually or in an automated way.
```

Comment:
Does the framework handle installation of dependencies (e.g. CLI tools)? Or do we anticipate handling that via `make` or similar?

Reply:
Also, is being able to manually execute some version of these tests using `kubectl` and other tools a goal here? I don't think it is a negative necessarily, but I also don't think manually going through the tests instill higher confidence in functionality, and I don't think we should strive for our e2e tests to also act as examples for end-users.

Reply:
> Does the framework handle installation of dependencies (e.g. CLI tools)? Or do we anticipate handling that via make or similar?

The only needed CLI is `kubectl` - if needed we could install it via `make`.

> is being able to manually execute some version of these tests using kubectl and other tools a goal here?

No, it is more just a statement that anybody reading the test could even manually reproduce it. A nice bonus point, but as you said, not a goal at all.

Reply:
+1 to e2es serving also as documentation being a nice to have, but not a requirement